activetable/activetable.py

```diff
@@ -0,0 +1,297 @@
+
+        frag = Fragment(html)
+        frag.add_css(css)
+        frag.add_javascript(loader.load_unicode('static/js/src/activetable.js'))
+        frag.initialize_js('ActiveTableXBlock', self.get_status())
+        return frag
+
+    def check_and_save_answers(self, data):
+        """Common implementation for the check and save handlers."""
+        if self.max_attempts and self.attempts >= self.max_attempts:
+            # The "Check" button is hidden when the maximum number of attempts has been reached, so
+            # we can only get here by manually crafted requests.  We simply return the current
+            # status without rechecking or storing the answers in that case.
+            return self.get_status()
+        self.parse_fields()
+        self.postprocess_table()
+        answers_correct = {
+            cell_id: self.response_cells[cell_id].check_response(value)
+            for cell_id, value in data.iteritems()
+        }
+        # Since the previous statement executed without error, the data is well-formed enough to be
+        # stored.  We now know it's a dictionary and all the keys are valid cell ids.
+        self.answers = data
+        return answers_correct
+
+    @XBlock.json_handler
+    def check_answers(self, data, unused_suffix=''):
+        """Check the answers given by the student.
+
+        This handler is called when the "Check" button is clicked.
+        """
+        self.answers_correct = self.check_and_save_answers(data)
+        self.attempts += 1
+        self.score = self.num_correct_answers * self.max_score / len(self.answers_correct)
+        self.runtime.publish(self, 'grade', dict(value=self.score, max_value=self.max_score))
+        return self.get_status()
+
+    @XBlock.json_handler
+    def save_answers(self, data, unused_suffix=''):
+        """Save the answers given by the student without checking them."""
+        self.check_and_save_answers(data)
+        self.answers_correct = None
+        return self.get_status()
+
+    def validate_field_data(self, validation, data):
+        """Validate the data entered by the user.
+
+        This handler is called when the "Save" button is clicked in Studio after editing the
+        properties of this XBlock.
+        """
+        def add_error(msg):
+            """Add a validation error."""
+            validation.add(ValidationMessage(ValidationMessage.ERROR, msg))
+        try:
+            thead, tbody = parse_table(data.content)
+        except ParseError as exc:
+            add_error('Problem with table definition: ' + exc.message)
+            thead = tbody = None
+        if data.column_widths:
+            try:
+                column_widths = parse_number_list(data.column_widths)
+            except ParseError as exc:
+                add_error('Problem with column widths: ' + exc.message)
+            else:
+                if thead is not None and len(column_widths) != len(thead):
```

Comment:
I intentionally didn't check it.  The 800px limit is a rule of thumb that works well when using this XBlock on edx-platform on desktop, but people might have other requirements anyway, so it's really up to them to decide whether it looks good enough for them.